File: 0046-permutations/0046-permutations.java
class Solution {
    public void backtrack_46(List<List<Integer>> itog, List<Integer> curr, boolean[] used, int[] nums) {
        if (nums.length == curr.size()) { 
            itog.add(new ArrayList<>(curr)); 
            return; 
        }

       
        for (int i = 0; i < nums.length; i++) {
            if (!used[i]) {
                curr.add(nums[i]);
                used[i] = true;
                backtrack_46(itog, curr, used, nums);
                used[i] = false;
                curr.remove(curr.size() - 1); 
            }
        }
    }
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> itog = new ArrayList<>();
        List<Integer> curr = new ArrayList<>();
        boolean[] used = new boolean[nums.length];
        backtrack_46(itog, curr, used, nums);
        return itog;
    }
}

File: 0046-permutations/0046-permutations.cpp
class Solution {
public:
    void backtrack_46(vector<int>& nums, vector<vector<int>>& itog, vector<bool>& used, vector<int>& curr) {
        if (curr.size() == nums.size()) {
            itog.push_back(curr);
            return;
        }

        for (int i = 0; i < nums.size(); i++) {
            if (!used[i]) {
                curr.push_back(nums[i]);
                used[i] = true;

                backtrack_46(nums, itog, used, curr);

                used[i] = false;
                curr.pop_back();
            }
        }
    }
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> itog;
        vector<bool> used(nums.size(), false);
        vector<int> curr;
        backtrack_46(nums, itog, used, curr);
        return itog;
    }
};